File: api/routes/leaderboard.py
"""
Leaderboard API endpoints.

GET /leaderboard               — default (streak-sorted) leaderboard
GET /leaderboard/streaks       — sorted by current streak
GET /leaderboard/consistency   — sorted by consistency %
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, Query

from api.middleware.cache import cached_route
from api.middleware.cache_headers import public_cache
from api.schemas.common import APIResponse
from api.schemas.leaderboard import LeaderboardEntry, LeaderboardResponse
from db import database

logger = logging.getLogger("dsa_bot.api.leaderboard")

router = APIRouter(prefix="/leaderboard", tags=["Leaderboard"])

# Valid sort keys matching existing leaderboard_handler.SORT_KEYS
_VALID_SORTS = {"streak", "longest", "consistency", "posts", "days"}
_SORT_KEY_MAP = {
    "streak": "current_streak",
    "longest": "longest_streak",
    "consistency": "consistency",
    "posts": "total_messages",
    "days": "days_posted",
}


async def _build_leaderboard(sort_by: str, limit: int = 25) -> LeaderboardResponse:
    """Shared leaderboard builder reusing database.get_leaderboard_data()."""
    result = await database.get_leaderboard_data()
    total_registered = result["total_registered_users"]
    data = result["rankings"]

    sort_key = _SORT_KEY_MAP.get(sort_by, "current_streak")
    data.sort(key=lambda x: x.get(sort_key, 0), reverse=True)

    active_streaks = len(data)  # every row already has streak >= 1
    entries = []

    for rank, entry in enumerate(data[:limit], start=1):
        entries.append(
            LeaderboardEntry(
                rank=rank,
                user_id=str(entry["user_id"]),
                discord_username=entry.get("discord_username"),
                username=entry.get("username"),
                current_streak=entry.get("current_streak", 0),
                longest_streak=entry.get("longest_streak", 0),
                consistency_pct=entry.get("consistency", 0.0),
                total_messages=entry.get("total_messages", 0),
                days_posted=entry.get("days_posted", 0),
            )
        )

    return LeaderboardResponse(
        sort_by=sort_by,
        total_users=total_registered,
        active_streaks=active_streaks,
        entries=entries,
    )


# ── Default ──────────────────────────────────────────────────────────────────

@cached_route(60)
@router.get("")
@router.get(
    "/",
    response_model=APIResponse[LeaderboardResponse],
    summary="Get leaderboard",
    description="Returns the leaderboard sorted by the specified criterion. "
                "Valid sort values: streak, longest, consistency, posts, days.",
)
async def get_leaderboard(
    sort_by: str = Query("streak", description="Sort criterion"),
    limit: int = Query(25, ge=1, le=100, description="Max entries to return"),
    _cache: None = Depends(public_cache),
):
    if sort_by not in _VALID_SORTS:
        sort_by = "streak"
    result = await _build_leaderboard(sort_by, limit)
    return APIResponse(data=result)


# ── Convenience Aliases ──────────────────────────────────────────────────────

@cached_route(60)
@router.get(
    "/streaks",
    response_model=APIResponse[LeaderboardResponse],
    summary="Leaderboard by streaks",
    description="Shortcut for leaderboard sorted by current streak.",
)
async def leaderboard_streaks(
    limit: int = Query(25, ge=1, le=100),
    _cache: None = Depends(public_cache),
):
    return APIResponse(data=await _build_leaderboard("streak", limit))


@cached_route(60)
@router.get(
    "/consistency",
    response_model=APIResponse[LeaderboardResponse],
    summary="Leaderboard by consistency",
    description="Shortcut for leaderboard sorted by consistency percentage.",
)
async def leaderboard_consistency(
    limit: int = Query(25, ge=1, le=100),
    _cache: None = Depends(public_cache),
):
    return APIResponse(data=await _build_leaderboard("consistency", limit))

File: mobile/lib/features/dashboard/data/models/leaderboard.dart
/// Immutable data models for the leaderboard API endpoint.
///
/// Endpoint: `GET /leaderboard?sort_by=streak&limit=25`
/// Response envelope: `APIResponse<LeaderboardData>`
library;

// =============================================================================
// LeaderboardEntry — a single ranked user row
// =============================================================================

class LeaderboardEntry {
  const LeaderboardEntry({
    required this.rank,
    required this.userId,
    this.discordUsername,
    this.username,
    this.currentStreak = 0,
    this.longestStreak = 0,
    this.consistencyPct = 0.0,
    this.totalMessages = 0,
    this.daysPosted = 0,
  });

  final int rank;
  final String userId;
  final String? discordUsername;
  final String? username;
  final int currentStreak;
  final int longestStreak;
  final double consistencyPct;
  final int totalMessages;
  final int daysPosted;

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) {
    return LeaderboardEntry(
      rank: (json['rank'] as num?)?.toInt() ?? 0,
      userId: json['user_id']?.toString() ?? '',
      discordUsername: json['discord_username']?.toString(),
      username: json['username']?.toString(),
      currentStreak: (json['current_streak'] as num?)?.toInt() ?? 0,
      longestStreak: (json['longest_streak'] as num?)?.toInt() ?? 0,
      consistencyPct: (json['consistency_pct'] as num?)?.toDouble() ?? 0.0,
      totalMessages: (json['total_messages'] as num?)?.toInt() ?? 0,
      daysPosted: (json['days_posted'] as num?)?.toInt() ?? 0,
    );
  }
}

// =============================================================================
// LeaderboardData — full response payload
// =============================================================================

class LeaderboardData {
  const LeaderboardData({
    this.sortBy = 'streak',
    this.totalUsers = 0,
    this.activeStreaks = 0,
    this.entries = const [],
  });

  final String sortBy;
  final int totalUsers;
  final int activeStreaks;
  final List<LeaderboardEntry> entries;

  factory LeaderboardData.fromJson(Map<String, dynamic> json) {
    final rawEntries = json['entries'] as List<dynamic>?;

    return LeaderboardData(
      sortBy: json['sort_by']?.toString() ?? 'streak',
      totalUsers: (json['total_users'] as num?)?.toInt() ?? 0,
      activeStreaks: (json['active_streaks'] as num?)?.toInt() ?? 0,
      entries: rawEntries
              ?.map((e) =>
                  LeaderboardEntry.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }
}
